### ISSUE TITLE
Docker image update policy

### ISSUE THREAD

**Issue by iamFIREcracker** [neo4j/docker-neo4j#342]
Hey,

My team is using a fairly old version of Neo4j (4.2.8 to be precise) and all of us got a bit surprised when we realized that the image that we had been using for almost a year, had [recently](https://hub.docker.com/_/neo4j?tab=tags&page=1&name=4.2.8) been updated:

> Last pushed 13 days ago by [doijanky](https://hub.docker.com/u/doijanky)

Even more surprising was that the newer version (or one that got "updated" sometime in between last summer and 13 days ago) was now shipping Debian 11, which unfortunately requires the host environment to run a more recent version of docker (see: https://github.com/neo4j/docker-neo4j/issues/318).  Note: the 4.2.8 version that we have, cached inside our CI/CD environment, runs Debian 10, and we never run into any permissions issue with it.

Now, we are working things out internally to try and switch to a newer version of docker on the host, and to a newer version of neo4j; however, I also wanted to understand more of how are the Neo4j docker images getting updated.  From where I stand it seems as if you were treating all the published tags as _rolling_ versions, i.e. like `latest`, where pulling it today or in one month might give different results; however, I question if this is what the different engineering teams are actually expecting, especially given that quite a big change (like the upgrade of the underlying OS) could be snuck into a tag rewrite and potentially set your whole CI/CD pipeline on fire.

I know we can refer to images by their digest, and that's what we will ultimately end up doing if we realized we could not trust a specific version, like `4.2.8`, to stay immutable; however, if version 4.2.8 today is different from version 4.2.8 from one week ago, how could I then (1) be notified about the update, and (2) audit the new version to confirm it's compatible with our deployment environment?

Thanks,
M.

**Comment by jennyowen** [comment]
Neo4j is a supported, official docker image. As such, the people at docker automatically rebuild our images when there are changes and updates to our base images.
At some point in the last year, debian updated their default debain slim to be bullseye rather than buster. This prompted an update of our base image `openjdk:11-jdk-slim`, which in turn caused Neo4j images to be rebuilt.

We maintain a list of supported Neo4j versions and their image tags for the people at docker:
https://github.com/docker-library/official-images/blob/master/library/neo4j
The listed images are rebuilt and updated whenever the base images are changed. This allows the Neo4j to get security updates as soon as they are available.

The parts of this update process we control are:

- Which Neo4j versions are in the supported list
- The contents of the Neo4j image that is NOT the openjdk base.

### Updating the Neo4j image contents
All the Neo4j image source code can be found here:
https://github.com/neo4j/docker-neo4j-publish

It is extremely rare that we change the contents of an image after publishing, and even then it will only be a change to the docker entrypoint scripts around Neo4j and **never** any part of the database itself.
When the docker image is built, it will download Neo4j from our server and this means the Neo4j docker image will always be running the same source code as a regular Neo4j installation. 

### Which Neo4j versions should get updates
Currently all supported major, minor and patch versions of Neo4j will automatically receive security updates from docker. 
The list of supported Neo4j versions and the support road map can be found at:
https://neo4j.com/developer/kb/neo4j-supported-versions/

When Neo4j 5.0.0 is released, we are only going to list the latest patch version of a major.minor version as supported in docker. This will apply retroactively to all existing Neo4j releases.
For example, of all the 4.2.X releases, only the latest patch will be marked to receive security updates in docker.

Neo4j versions not listed at https://github.com/docker-library/official-images/blob/master/library/neo4j are still available for use, they just do not get the updates to the base image.

**Comment by jennyowen** [comment]
For your question about getting updates about when an image is updated, I don't know. These updates are done by docker, not Neo4j, you would have to ask docker support about how to monitor a particular tag for changes.

**Comment by jennyowen** [comment]
> From where I stand it seems as if you were treating all the published tags as rolling versions, i.e. like latest, where pulling it today or in one month might give different results;

Yes I suppose that is true. Although, I would argue that this is true for **all** docker tags. Docker tags are mutable by design. Unless you manage the image yourself (or specify which image you want by digest) you do risk unexpected changes.

For `neo4j:4.2.8`, the promise is that it will ALWAYS have only neo4j version 4.2.8, there will never be forced updates of Neo4j, but the various undocumented parts of the image that support neo4j are indeed subject to change.
